core/backend/src/IModelDb.ts

```diff
@@ -2388,8 +2388,8 @@ export class BriefcaseDb extends IModelDb {
     // good thing computers are fast. Fortunately upgrading should be rare (and the push time will dominate anyway.) Don't try to optimize any of this away.
     await withBriefcaseDb(briefcase, async (db) => db.acquireSchemaLock()); // may not really acquire lock if iModel uses "noLocks" mode.
     try {
-      await this.doUpgrade(briefcase, { profile: ProfileOptions.Upgrade }, "Upgraded profile");
-      await this.doUpgrade(briefcase, { domain: DomainOptions.Upgrade }, "Upgraded domain schemas");
+      await this.doUpgrade(briefcase, { profile: ProfileOptions.Upgrade, schemaLockHeld: true }, "Upgraded profile");
```

Comment:
is all this still necessary (opening/closing/pushing changes in separate changesets?)

Reply:
Yes. We need to refactor this. We have overloaded open() call with too many operations. It should just open the file. Domain/Profile flags are now same as domain will upgrade profile if needed. Which mean we just need two methods

1. Db::Open() 
2. Db::Upgrade().

If the open fails to open, then you call upgrade it just upgrades too latest (profile and schemas).